### PR TITLE
Fix several bugs in TextBox

### LINKED ISSUE #493: Pressing Ctrl-Z in a TextBox with no changes removes text
It should do nothing.

### PR DESCRIPTION
**Pull request by andrewst**
Fixed #493.

### CODE REVIEW

File: src/Avalonia.Controls/Utils/UndoRedoHelper.cs [hunk 1]
```diff
@@ -8,44 +8,41 @@
...
 
         public UndoRedoHelper(IUndoRedoHost host)
         {
             _host = host;
-            _states.AddFirst(_host.UndoRedoState);
-            _currentNode = _states.First;
-            WeakTimer.StartWeakTimer(this, new TimeSpan(0, 0, 1));
-
+            //_states.AddFirst(_host.UndoRedoState);
+            //_currentNode = _states.First;
```
Comment (grokys) [inline src/Avalonia.Controls/Utils/UndoRedoHelper.cs]: Could you remove the commented-out code? I don't like commented-out in general as it's difficult to tell why it was commented out when we come back to it in a year's time.

**Comment by kekekeks** [comment]
Why exactly are you changing state to struct?

**Comment by andrewst** [comment]
Because "_currentNode" in the current implementation may be null.
"LastState" property was returned "Nullable\<TState\>".
Nullable can only work with the structures.
https://github.com/AvaloniaUI/Avalonia/pull/733/commits/e3f94011bc8f82241bcf661bae374428b87a0d75#diff-3c96d2967cc4d98d4f60a297659fd3c5R57

**Comment by kekekeks** [comment]
1) I don't like the idea of limiting UndoRedoHelper to structs.
2) I don't like that you are checking for `_undoRedoHelper.LastState?.Text == Text`, it's not obvious what it does. It's true if last_state == Text || (there were no state && Text == null). It's better to introduce `HasState` property and work with that.


BTW, the bug isn't fixed. Because it's not "Pressing Ctrl-Z in a TextBox with no changes removes text" it's "When TextBox Text is changed through code or bindings undo history is not reset".

So the UndoRedoManager was OK, one needs to fix TextBox instead to properly recognize the source of changes of `Text` property and reset undo history when needed.

**Comment by kekekeks** [comment]
I'd recommend to not touch UndoRedoHelper at all (may be add `Clear` method that discard all saved states and saves the current one).

**Comment by kekekeks** [comment]
Oh, I've checked WPF, it saves states from code/bindings. Point about `HasState` still stands, otherwise seems ok.

**Comment by andrewst** [comment]
Struct is good, becouse, when we return LastState our inner state nobody cann't change from outer.

1. HasState really be better.
2. But then I propose to replace property LastState to bool TryGetLastState(out TState _State).
We can remove restrictions struct, but I think the structure will be better.

**Comment by andrewst** [comment]
I made the necessary changes.
What did you decide on this request?

**Comment by kekekeks** [comment]
Please, add a unit test for the issue and I think we shall merge it

File: tests/Avalonia.Controls.UnitTests/TextBoxTests.cs [hunk 2]
```diff
@@ -42,6 +42,58 @@ public void CaretIndex_Can_Moved_To_Position_After_The_End_Of_Text_With_Arrow_Ke
...
+
+                RaiseKeyEvent(target, Key.A, InputModifiers.Control);
+
+                Assert.Equal(0, target.SelectionStart);
+                Assert.Equal(0, target.SelectionEnd);
+            }
+        }
+
+        [Fact]
+        public void Press_Ctrl_Z_With_Not_Modify_Text()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/TextBoxTests.cs]: I think that should be "Will", not "With" ;)

**Comment by grokys** [comment]
LGTM except the typo. @kekekeks are we OK to merge?

**Comment by kekekeks** [comment]
@andrewst Please, update your PR to the latest state of `master`, github won't automatically merge otherwise.

**Comment by grokys** [comment]
@andrewst sorry, I've just got round to trying to merge this now in the absense of @kekekeks and your PR is out of date with master again :(

Could you merge again, or if you could give me [write access to the PR branch](https://github.com/blog/2247-improving-collaboration-with-forks) I could merge myself.

**Comment by grokys** [comment]
Thanks @andrewst - merged finally! :)

**Pull request merged**